### PR TITLE
Drain the pipeline before forcing it to NULL

### PR DESCRIPTION
**Pull request by pksgit**
## The problem

`SendEOS`, despite its name, sent no EOS. It called `BlockSetState(StateNull)` immediately, while media was still flowing. On the kill path that is a WHIP `DELETE`, which ffmpeg sends at the end of every normal session.

On GStreamer 1.28.6 that state change collides with a streaming thread mid-chain:

```
Thread 4 (SendEOS goroutine)        Thread 32 "multiqueue1:src"
  gst_element_set_state(NULL)         gst_base_transform_chain()  <- holds transform lock
    gst_pad_set_active(FALSE)           gst_video_rate_transform_ip()
      gst_base_transform_activate()       do { flush_prev(); } while (diff1 < diff2)
        pthread_mutex_lock()  <- blocks     <- runs away: out 1,183 -> 93,493,506
```

Captured mid-stall, `videorate` was the only element with an unfinished transition, across nine consecutive samples, while its output counter climbed at roughly 21 million frames per second with no input arriving. Teardown then waits out the full 5 second fallback and logs `pipeline frozen`.

## The change

Send EOS and let the existing bus watch stop the pipeline once the event reaches the bus, so no chain call is in flight when the state change runs. Three exits, each with its own sequence:

- **not running** (below PAUSED): force NULL and quit. EOS needs streaming threads to travel, so there is nothing to drain. This runs off the caller's goroutine so a slow state change cannot block the kill path.
- **drained**: quit, since `messageWatch` already set NULL.
- **timed out**: force NULL, log, quit.

The 5 second timer keeps its old job as a fallback, and now forces NULL before logging, so the worst case is unchanged rather than a new hang.

`drained` is a `core.Fuse` to match `closed` beside it. A plain channel would need `New` to initialise it, and the tests construct `Pipeline` directly, so `close(nil)` panics. A fuse is zero-value safe.

## Evidence

Interleaved A/B trials, 40 second transcoding WHIP sessions, arms alternated so machine drift cannot explain the gap.

| Trial | Stock | With fix | Significance |
|---|---|---|---|
| 1.28.6 efficacy, pooled over 2 runs | 17 / 50 | 1 / 50 | p = 0.000033 |
| 1.26.7 compatibility (ships today) | 0 / 20 | 0 / 20 | no difference |
| 1.28.6 vs 1.26.7, the bug itself | 7 / 20 | 0 / 20 | p = 0.0083 |

Two independent results corroborate: 1.26.7 never reproduces the stall, and egress is unaffected because it already drains via `AddEOSProbe` before calling `Stop()`, running 15 of 15 clean on the same 1.28.6 image with the same `videorate` element in its pipelines.

## Test

`TestSendEOSDrainsARunningPipeline` asserts on the drain rather than on the stop, since the drain is what changed. Verified by breaking the code:

| Code under test | Result | Caught by |
|---|---|---|
| The fix as written | PASS 0.05s | - |
| `SendEOS` reverted to the stock force-NULL | FAIL 0.03s | `drained.IsBroken()` |
| `SendEvent(NewEOSEvent())` removed | FAIL 3.01s | the 3 second bound |

The second row is the point. Stock still returns in about 50ms, because forcing `StateNull` is fast, so a timing-only test would have passed against the bug and guarded nothing.

## Why this is not shared with egress

Egress already solves this, and its solution is not liftable:

- Egress drains per bin through `AddEOSProbe`, tied to its own `gstreamer.Bin` wrapper: a Go-side tree with per-node state management, callbacks and per-peer EOS accounting, walked recursively so EOS is injected at each leaf source and confirmed at each sink, under a 30 second `eosTimeout`. Ingress nests real GStreamer bins but has none of that machinery, so its drain is one pipeline-level EOS and one bus message.
- The two live in separate repos with no shared module. A common helper means a third repo or a protocol-side package, for roughly 30 lines.
- The lifecycles genuinely differ. Egress tears down on a request that can wait 30 seconds; ingress tears down on a publisher disconnect where the source is already gone.

A shared abstraction here would be wider than either caller needs.

## Known limits

- One session in fifty still froze. That run did build a pipeline, so it is not the not-running case the guard covers.
- This prevents the trigger; it cannot rescue a running loop. EOS is a serialized event, so if `videorate` is already spinning, EOS queues behind the stuck chain call and the fallback fires: same outcome as today, five seconds later.
- If a `KillIngress` races a pipeline error, teardown takes 5 seconds where it was immediate. The `MessageError` branch removes the bus watch without setting NULL, so the EOS has nowhere to go and the fallback handles it. Bounded and logged, though `pipeline frozen` is misleading in that case. Fixing it properly means setting NULL in the error branch, which is a change to the error path beyond this PR.
- The root cause is still open. Both pads were flushing throughout the stall, so `flush_prev` should have returned non-OK and broken the loop. Why it does not is unexplained, and no input should make `videorate` emit 93 million frames from two. This is a mitigation, not the upstream fix.

## Verification

Against GStreamer 1.26.11, the version in `.gst-version`, using the exact CI checks:

- `golangci-lint run --timeout=5m --modules-download-mode=mod`: pass
- `go test -v -timeout 20m ./pkg/...`: 8 packages ok, 37 pass, 0 fail

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by pksgit** [comment]
Looking into the integ failures and above comments so converting this to draft mode for now.

File: pkg/media/pipeline.go [hunk 1]
```diff
@@ -492,16 +494,26 @@ func (p *Pipeline) SendEOS(ctx context.Context) {
...
 	}
 
-	c := make(chan struct{})
+	// Draining only works once the pipeline is running: EOS has to travel the
+	// streaming threads to reach the bus. Below PAUSED nothing carries it, so
+	// the drain would always time out. Stop directly in that case.
+	// The state read can be stale if Run is mid-startup, but that window is the
+	// same one the unconditional stop already had. Stop off the caller's
+	// goroutine so a slow state change never blocks the kill path.
+	if p.pipeline.GetCurrentState() < gst.StatePaused {
```
Comment (biglittlebigben) [inline pkg/media/pipeline.go:503]: This is probably racy

File: pkg/media/pipeline.go [hunk 2]
```diff
@@ -492,16 +494,26 @@ func (p *Pipeline) SendEOS(ctx context.Context) {
...
 
-	go func() {
-		err := p.pipeline.BlockSetState(gst.StateNull)
-		if err != nil {
-			logger.Errorw("failed stopping pipeline", err)
-		}
+	// Drain first: EOS travels the streaming threads and messageWatch stops the
+	// pipeline once it reaches the bus, so no chain call is in flight when the
+	// state change runs. Going straight to NULL races those threads.
+	p.pipeline.SendEvent(gst.NewEOSEvent())
```
[the remark this answers is not in the document]
Comment (biglittlebigben) [reply pkg/media/pipeline.go:514]: Yes, we should start the timer earlier

File: pkg/media/pipeline.go [hunk 3]
```diff
@@ -510,6 +522,7 @@ func (p *Pipeline) SendEOS(ctx context.Context) {
 		case <-c:
 			t.Stop()
 		case <-t.C:
+			p.forceStop()
```
[the remark this answers is not in the document]
Comment (biglittlebigben) [reply pkg/media/pipeline.go:525]: I believe Devin is right here

**Comment by pksgit** [comment]
Closing this in favour of a different fix: the root cause turned out to be in ingress, not in teardown.

`videorate` was livelocking because it received buffers with an unusable timestamp, and teardown only made that visible: `set_state(NULL)` blocked on the lock the spinning chain call held, the 5 second fallback fired, and the session logged `pipeline frozen`.

The timestamp came from us. In `WHIPSource.getCorrctedTimestamp` both tracks share one `startOffset`, latched by whichever relay delivers its first packet first. The other track can legitimately carry an earlier timestamp, so the subtraction goes negative, and `GstClockTime` is unsigned, so a negative `time.Duration` reaches the pipeline as a value near 2^64.

Captured end to end in one session on 1.28.6:

```
audio first packet raw = 117,835 ns  -> latches startOffset, corrected = 0
video first packet raw = 1 ns        -> corrected = -117,834 ns
gst.ClockTime(-117834)               -> 18,446,744,073,709,433,782
videorate logs                          "got buffer with timestamp 5124095:34:33.709433782"
```

which is the same number. The negative only matters when it lands on the video track, since `videorate` exists only in the video bin. Over 25 interleaved sessions on stock, video negatives and freezes matched exactly, 8 and 8.

A one-line clamp in `getCorrctedTimestamp` takes that from 8 in 25 to 0 in 25 (Fisher exact, two-sided, p = 0.004).

Two things from this PR that are worth keeping, and I would rather raise them separately than bury them here:

- The review findings on the EOS restructuring were correct and I confirmed both: a potentially blocking send must not sit on the caller's goroutine ahead of the fallback being armed, and `forceStop` must not sit on the path to `quitLoop`. The current teardown code is unchanged and still has those properties, so nothing is broken today, but the reasoning is worth preserving.
- `videorate` livelocking rather than rejecting an unusable timestamp is an upstream defect regardless of what we feed it.
